src/metabase/query_processor/streaming/xlsx.clj

```diff
@@ -17,38 +19,68 @@
    :headers                   {"Content-Disposition" (format "attachment; filename=\"query_result_%s.xlsx\""
                                                              (u.date/format (t/zoned-date-time)))}})
 
-(defprotocol ExcelFormatValue
-  "Protocol for specifying how objects of various classes in QP result rows should be formatted for Excel
-  exports. The generic version in `common/FormatValue` does not work for Excel due to date/time formats."
-  (excel-format-value [this]
-    "Format this value in a QP result row appropriately for a results download to Excel."))
+;; docjure has an open issue about its formatting issues: https://github.com/mjul/docjure/issues/75
+;; since this hasn't been addressed at the library level, we have to do some workarounds
 
-;; this version of the protocol resolves #10803 by not converting datetimes to ISO8601 (which Excel cannot parse)
-;; Docjure can handle java.util.Date instances, but not the new Java 8 Time instances, so we convert everything
-;; to java.util.Date.
-(extend-protocol ExcelFormatValue
-  nil
-  (excel-format-value [_] nil)
+;; this method is private in the docjure library, but we need to able to call it from here
+;; also, the version in Docjure doesn't keep track of created cell styles, which is bad
+;; these need to be memoized per Workbook
 
-  Object
-  (excel-format-value [this] this)
+(def ^:dynamic *cell-styles*
+  "Holds the CellStyle values used within a spreadsheet so that they can be reused. Excel has a limit
+   of 64,000 cell styles in a single workbook."
+  nil)
 
-  java.time.temporal.Temporal
-  (excel-format-value [this]
-    (t/java-date this))
+(defn- create-or-get-date-format [^Workbook workbook ^String format-string]
+  (when-not (contains? @*cell-styles* format-string)
+    (let [new-val (let [format-helper (.getCreationHelper workbook)
+                        date-style (.createCellStyle workbook)]
+                    (assoc
+                     @*cell-styles*
+                     format-string
+                     (doto date-style
+                       (.setDataFormat (.. format-helper createDataFormat (getFormat format-string))))))]
+      (reset! *cell-styles* new-val)))
```

Comment:
why `reset!` instead of `swap!` + `assoc`?

Reply:
:shrug: Is one style preferred over the other?

Reply:
`swap!` is preferable because it performs a transformation atomically while `reset!` replaces the whole thing atomically. It doesn't really matter here because it's non running on multiple threads but `swap!` will do all the updates without stomping on changes from other threads while `reset!` will just completely replace the contents of the atom with no regards to what any other threads might be doing

Reply:
swap! = apply a transformation
reset! = completely reset the contents of the atom